### ISSUE TITLE
BIP Firmware from Mi Fit 4.0

### ISSUE THREAD

**Issue by grzegorzlada** [Freeyourgadget/Gadgetbridge#1514]
Something changed in firmware distribution since Mi Fit 4.0 and Amazfit BIP A1608. Mi Fit 4.0 (variants 7280-7282) comes with embedded Mili_chaohu.fw V1.1.5.16, however if you run the Mi Fit it tries to update BIP's firmware to V 1.1.5.23. How can you acquire new FW file?

**Comment by ashimokawa** [comment]
You probably would need a rooted  phone, and while doing the update copy the temporary firmware file to some safe location before mifit deletes it.

This is only guessing - never used mi fit.

**Comment by ilyasovivan** [comment]
1.1.5.23 - it was a beta version. It is now stable at 1.1.5.36 (.latin version) - for some languages and 1.1.5.24 (not .latin) - for English and Chinese. Today began the update on all devices using MiFit 4.0.0 (online FW update).

**Comment by ashimokawa** [comment]
Damn, so OTA updates are back. Thanks for reporting!

**Comment by grzegorzlada** [comment]
It looked like OTA update as it was trying to perform an update after ~3rd application launch. I was looking for the downloaded firmware in data directory on not-rooted phone but without any luck.
Is there any workaround to get the .fw file?

BTW - what do you mean they are back? Did it happen before?

**Comment by ilyasovivan** [comment]
> Is there any workaround to get the .fw file?
https://myamazfit.ru/threads/bip-oficialnaja-proshivka-1-1-5-24-1-1-5-36-iz-mifit-4-0-0.1005

**Comment by ashimokawa** [comment]
@grzegorzlada 
Yes, in the early days. 0.0.8.96-98, three versions were OTA only,

@ilyasovivan 
Thanks for the link, unfortunately downloads require registration, but better then nothing ;)

**Comment by grzegorzlada** [comment]
Yeah, thank you very much for the download link - amazing community here! :)

**Comment by ilyasovivan** [comment]
> Thanks for the link, unfortunately downloads require registration

Oh, sry. Direct link:
https://cloud.mail.ru/public/32iG/jSSXCagog
Not Latin (english, chinese):
FW ver: 1.1.5.24
RES ver: 61
Font ver: 8 (not new)
GPS ver: 15974,e61dd16,126 (not new)

Latin (Spanish, Russian and new languages: French, German, Italian, Turkish):
FW ver: 1.1.5.36
RES ver: 61 (the same)
Font ver: 9 (not new)
GPS ver: 15974,e61dd16,126 (not new, the same) - previous name "mili_dth.gps"

**Comment by ashimokawa** [comment]
Flashing the Firmware works without flashing the res. When flashing the res, the watch asks me to update, which usually means the res does not fit the firmware.

**Comment by grzegorzlada** [comment]
Right, so what should we do about this issue?
- Close it as the current firmware was linked here, or
- Leave it open to figure out a mechanism for obtaining OTA firmware updates?

**Comment by elokheizer** [comment]
I installed 1.1.5.36 via MiFit.
But I can't select German language in Gadgetbridge.
Now my Bip is Spanish. Time for learning a new language? ;-)

**Comment by ashimokawa** [comment]
@elokheizer: Setting your phone to German and Bip language to automatic in Gadgetbridge settings should work iirc. Or compile master ;)

**Comment by DerFetzer** [comment]
When I remember correctly and look at the source, setting the language to automatic would not work.
At the moment in `AmazfitBipSupport.java` it is:
```java
switch (language) {
...
case "es":
        command_old = AmazfitBipService.COMMAND_SET_LANGUAGE_SPANISH;
        localeString = "es_ES";
        break;
...
    default:
        command_old = AmazfitBipService.COMMAND_SET_LANGUAGE_ENGLISH;
        localeString = "en_US";
        break;
}
```
So it should default to English. And when you send a language code to the watch that is not available in the current firmware then it falls back to the first language which should be Spanish for Latin.

You could do something like:
```java
switch (language) {
...
    default:
        command_old = AmazfitBipService.COMMAND_SET_LANGUAGE_ENGLISH;
            localeString = language + "_" + country;
            break;
}
```
I can test it if you think changing the default behavior at that point is a good idea.

**Comment by ashimokawa** [comment]
@DerFetzer 
Sorry. Mi Band 3 was like I described, not Bip/Cor.
I refactored a lot of code and changed it to match the behavior of Mi Band 3, so basically what you suggested above

**Comment by elokheizer** [comment]
Now my bip is German. Thank you for new Gadgetbridge update.

**Comment by DerFetzer** [comment]
@ashimokawa  Nothing to be sorry about! Thank you for your great work.

**Comment by ashimokawa** [comment]
Latest Mi Fit 4.0.1 dropped all included fimware files except for Mi Bands, Shoes and Scales.
:scream:

**Comment by b16r05** [comment]
For Amazfit **Cor**, the last fw I have installed is FW:1.2.7.22; RES: 69 from MiFit com.xiaomi.hm.health_**3.6.0**-7270_minAPI19(armeabi-v7a)(nodpi)_apkmirror.com.apk

With the latest MiFit apk, com.xiaomi.hm.health_**4.0.6**-7778_minAPI19(armeabi-v7a)(nodpi)_apkmirror.com.apk, there is no 

Mili_tempo.fw
Mili_tempo.res 

Does this mean no more fw updates for Amazfit Cor? Or, have the file names changed?

**Comment by spaetz** [comment]
The latest MiFit 4.0.8 app also has nothing like Mili_chaohu* in its assets folder. ..

**Comment by ilyasovivan** [comment]
> The latest MiFit 4.0.8 app also has nothing like Mili_chaohu* in its assets folder. ..

And? Now all the firmware update via OTA.

**Comment by spaetz** [comment]
"And?" Well, that is NOT documented in the wiki page on how to update the Mili_chaohu firmware page at all, so I did not know that.
If that is the case it should be mentioned there (I can do that if that is indeed the case).

**Comment by vanous** [comment]
The Wiki did mention OTA but i have now added more explicit mention of that. https://codeberg.org/Freeyourgadget/Gadgetbridge/wiki/Amazfit-Bip-Firmware-Update

**Comment by spaetz** [comment]
Thanks a lot, much clearer now 😃.
-- 
Sent from mobile device. Please excuse brevity and tpyos.
